File: android_frontend/app/src/main/java/com/tracknrace/android/NativeTelemetry.java
package com.tracknrace.android;

final class NativeTelemetry {
    interface Listener {
        void onTelemetryRow(String json);
    }

    static {
        System.loadLibrary("tracknrace_android");
    }

    private final Listener listener;

    NativeTelemetry(Listener listener) {
        this.listener = listener;
    }

    String start(int port) {
        return nativeStart(port);
    }

    void stop() {
        nativeStop();
    }

    void setRecording(boolean enabled, String outputDirectory) {
        nativeSetRecording(enabled, outputDirectory);
    }

    @SuppressWarnings("unused") // Invoked by JNI on libtnrp worker threads.
    private void onNativeRow(String json) {
        listener.onTelemetryRow(json);
    }

    private native String nativeStart(int port);
    private native void nativeStop();
    private native void nativeSetRecording(boolean enabled, String outputDirectory);
}

File: android_frontend/app/src/main/java/com/tracknrace/android/NativeTelemetry.kt
package com.tracknrace.android

internal class NativeTelemetry(
    private val listener: Listener,
) {
    interface Listener {
        fun onTelemetryRow(json: String)

        fun onTelemetryBinary(bytes: ByteArray)
    }

    companion object {
        init {
            System.loadLibrary("tracknrace_android")
        }
    }

    fun start(port: Int): String? = nativeStart(port)

    fun stop() = nativeStop()

    fun setRecording(enabled: Boolean, outputDirectory: String) {
        nativeSetRecording(enabled, outputDirectory)
    }

    @Suppress("unused") // Invoked by JNI on libtnrp worker threads.
    private fun onNativeRow(json: String) {
        listener.onTelemetryRow(json)
    }

    @Suppress("unused") // Invoked by JNI on libtnrp worker threads.
    private fun onNativeBinary(bytes: ByteArray) {
        listener.onTelemetryBinary(bytes)
    }

    private external fun nativeStart(port: Int): String?

    private external fun nativeStop()

    private external fun nativeSetRecording(enabled: Boolean, outputDirectory: String)
}
